Reduce serialization errors in MultiWriterIdGen

erikjohnston opened the pull request:
We call `_update_stream_positions_table_txn` a lot, which is an UPSERT
that can conflict in `REPEATABLE READ` isolation level. Instead of doing
a transaction consisting of a single query we may as well run it outside
of a transaction.

clokep commented on synapse/storage/util/id_generators.py:
```diff
@@ -570,6 +574,19 @@ def _update_stream_positions_table_txn(self, txn):
+            with conn.cursor(txn_name="MultiWriterIdGenerator._update_table") as cur:
+                self._update_stream_positions_table_txn(cur)
+        finally:
+            conn.conn.autocommit = False  # type: ignore
```
Should we set this back to what it was instead of always `False`?

erikjohnston replied:
I mean, it really should always be `False`, but can do if you think its clearer?

richvdh commented on synapse/storage/util/id_generators.py:
```diff
@@ -570,6 +574,19 @@ def _update_stream_positions_table_txn(self, txn):
+    def _update_stream_positions_table_conn(self, conn: LoggingDatabaseConnection):
+        # We use autocommit here so that we don't have to go through a
+        # transaction dance, which a) adds latency and b) runs the risk of
+        # serialization errors.
```
don't we still risk serialization errors, but now they won't be handled?

autocommit is just syntactic sugar for a transaction with explicit commit afaik?

erikjohnston replied:
I believe it'll behave (at worst) like it was in a transaction with a `READ COMMITTED` isolation level, which doesn't have `concurrent update` errors. (Plus the fact that the command gets committed immediately means the likelihood of concurrent updates is dramatically smaller.


> ... but now they won't be handled?

Ah, I forgot that we handled those exceptions in the `new_transaction` function, rather than in `runWithConnection`. I do think that we'll still get errors even if we're not in a transaction though?

erikjohnston replied:
Of course having said all that the CI is failing with:

```
psycopg2.errors.SerializationFailure: could not serialize access due to concurrent update
```

erikjohnston commented:
Ok, after quite a bit of faff I think this is ready again. I changed it lots, which is why I rebased it into nicer commits.

The reason everything went horribly wrong is because we put newly connected connections into the DB pool with uncommitted transactions in them (whoops), which then meant nothing of the fiddling of isolation levels/auto commit modes worked. (Plus if you tried to do `conn.autocommit = True` the twisted DB connection wrapper doesn't proxy that to the underlying connection; it has implemented `__getattr_` proxy but not `__setattr__` proxy: https://twistedmatrix.com/trac/ticket/9998#ticket.) 

The current changes makes it possible to run queries outside of transactions, which should be safe for this use case (and I believe more generally for other single query transactions). Errors are handled, though we don't retry the queries again, which is fine here as we'll right to the table again soon (and the start up code handles it being behind).

I've manually tested concurrently executing these queries with a `READ COMMITTED` isolation level, and that works correctly (the second query blocks until the other transaction commits/rolls back).

The reason I have added the `db_autocommit` to `runInteraction`/`runWithConncetion` as I think we have a bunch of upserts that it should be safe to change to using this, which should hopefully reduce the amount of serialization errors we see.

richvdh commented on synapse/storage/database.py:
```diff
@@ -461,8 +461,23 @@ def new_transaction(
+            exception_callbacks
+            func
+            *args
+            db_retry: Whether to retry the transaction by calling `func` again.
+                This should be disabled if connection is in autocommit mode.
```
```suggestion
            db_retry: Whether to retry the transaction after an OperationalError by calling `func` again.
                This should be disabled if connection is in autocommit mode.
```

richvdh replied:
the problem with disabling this is that it's the thing that deals with postgres connections dropping (which can happen for annoying TCP reasons). I'm not sure it's safe to do: why is it required?

erikjohnston replied:
Unlike functions that are run in transactions its not necessarily true its safe to just re-run transaction functions, since they might have done half the work already. I guess we can mandate that such functions should be safe to re-run? (All the usages are for example)

erikjohnston replied:
I've done that.

richvdh commented on synapse/storage/engines/sqlite.py:
```diff
@@ -105,6 +107,14 @@ def server_version(self):
+        return conn.in_transaction  # type: ignore
+
+    def set_autocommit(self, conn: Connection, autocommit: bool):
+        # Twisted doesn't let us set attributes on the connections, so we can't
+        # set the connection to autocommit mode.
```
what's Twisted to do with it?

erikjohnston replied:
The connections we get are Twisted connections that wrap the underlying native connections, the wrapper implements `__getattr__` but not `__setattr__` so we can't set the autocommit flag

richvdh replied:
oh, right.

I suppose you could delve into `conn._connection`, but I guess it doesn't really matter? It might be worth commenting somewhere (at the call site, maybe), that this is will only *attempt* to enable autocommit, so the transaction must do an explicit commit anyway?

erikjohnston replied:
Yeah, we could go and pull out `_connection`, but then I'd want to handle the case where they rename that variable in a future release.

richvdh commented on synapse/storage/util/id_generators.py:
```diff
@@ -548,7 +548,7 @@ def _add_persisted_position(self, new_id: int):
                 break
 
-    def _update_stream_positions_table_txn(self, txn):
+    def _update_stream_positions_table_txn(self, txn: LoggingTransaction):
```
I'd recommend annotating this as `synapse.storage.types.Cursor` rather than `LoggingTransaction` unless you're actually relying on the extensions that are only in `LoggingTransaction`.

erikjohnston commented on synapse/storage/engines/postgres.py:
```diff
@@ -107,6 +108,7 @@ def on_new_connection(self, db_conn):
         db_conn.set_isolation_level(
             self.module.extensions.ISOLATION_LEVEL_REPEATABLE_READ
         )
+        db_conn.set_session(self.module.extensions.ISOLATION_LEVEL_REPEATABLE_READ)
```
Empirically it sets the postgres default isolation level, but it turns out to be pointless and I should remove it.

richvdh commented on synapse/storage/database.py:
```diff
@@ -600,7 +603,9 @@ async def runInteraction(
                 i.e. outside of a transaction. This is useful for transaction
```
```suggestion
                i.e. outside of a transaction. This is useful for transactions
```

richvdh commented on synapse/storage/database.py:
```diff
@@ -600,7 +603,9 @@ async def runInteraction(
                 i.e. outside of a transaction. This is useful for transaction
                 that are only a single query. Currently only affects postgres.
```
I think we could expand on this.

```suggestion
                that are only a single query. 
                
                Currently, this is only implemented for Postgres. SQLite will still
                run the function inside a transaction.
                
```

richvdh reviewed (approved):
lgtm otherwise
